ext/curl/url.c

```diff
@@ -0,0 +1,196 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Author: Pierrick Charron <pierrick@php.net>                          |
+   +----------------------------------------------------------------------+
+*/
+
+#define ZEND_INCLUDE_FULL_WINDOWS_HEADERS
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "php.h"
+#include "Zend/zend_exceptions.h"
+
+#include "curl_private.h"
+
+#if LIBCURL_VERSION_NUM >= 0x073E00 /* Available since 7.62.0 */
+
+/* CurlUrl class */
+zend_class_entry *curl_CURLUrl_ce;
+zend_class_entry *curl_CURLUrlException_ce;
+
+static zend_result php_curl_url_set(php_curlurl *uh, zend_long option, const char *str, const size_t len, unsigned int flags)
+{
+	CURLUcode error = curl_url_set(uh->url, option, str, flags);
+	SAVE_CURL_ERROR(uh, error);
+
+	return error == CURLUE_OK ? SUCCESS : FAILURE;
+}
+
+static void php_curl_url_ctor(INTERNAL_FUNCTION_PARAMETERS)
+{
+	php_curlurl *uh;
+	zend_string *url = NULL;
+
+	ZEND_PARSE_PARAMETERS_START(0,1)
+		Z_PARAM_OPTIONAL
+		Z_PARAM_STR_OR_NULL(url)
+	ZEND_PARSE_PARAMETERS_END();
+
+	if (NULL != url && strlen(ZSTR_VAL(url)) != ZSTR_LEN(url)) {
+		zend_argument_error(curl_CURLUrlException_ce, 1, "must not contain any null bytes");
+		RETURN_THROWS();
+	}
+
+	uh = Z_CURL_URL_P(return_value);
+	uh->url = curl_url();
+
+	if (url) {
+		if (php_curl_url_set(uh, CURLUPART_URL, ZSTR_VAL(url), ZSTR_LEN(url), 0) == FAILURE) {
+#if LIBCURL_VERSION_NUM >= 0x075000 /* Available since 7.80.0 */
+			zend_throw_exception(curl_CURLUrlException_ce, curl_url_strerror(uh->err.no), uh->err.no);
+#else
+			zend_throw_exception(curl_CURLUrlException_ce, "Unable to create CurlUrl", uh->err.no);
+#endif
+			RETURN_THROWS();
+		}
+	}
+}
+
+PHP_METHOD(CurlUrl, __construct)
+{
+	return_value = ZEND_THIS;
+	php_curl_url_ctor(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+}
+
+PHP_METHOD(CurlUrl, set)
+{
+	zval *zid;
+	zend_long part, flags = 0;
+	php_curlurl *uh;
+	zend_string *content;
+
+	if (zend_parse_method_parameters(ZEND_NUM_ARGS(), getThis(), "OlS|l", &zid, curl_CURLUrl_ce, &part, &content, &flags) == FAILURE) {
```

Comment:
Why is the old ZPP used here and the new one in the constructor?

Reply:
It was artefacts from when the implementation was OO and procedural. Did the change. Thanks